File: packages/sdk-go/webhooks_test.go
package governance_test

import (
	"testing"

	governance "github.com/org/platform/sdk-go"
)

func TestVerifySignature_valid(t *testing.T) {
	payload := []byte(`{"event":"policy.created"}`)
	secret := "test-secret-abc123"

	// Compute expected signature.
	import_crypto_hmac "crypto/hmac"
	import_sha256 "crypto/sha256"
	import_hex "encoding/hex"
	mac := import_crypto_hmac.New(import_sha256.New, []byte(secret))
	mac.Write(payload)
	sig := "sha256=" + import_hex.EncodeToString(mac.Sum(nil))

	if err := governance.VerifySignature(payload, sig, secret); err != nil {
		t.Fatalf("expected valid signature, got: %v", err)
	}
}

func TestVerifySignature_invalid(t *testing.T) {
	payload := []byte(`{"event":"policy.created"}`)
	if err := governance.VerifySignature(payload, "sha256=badhex", "secret"); err == nil {
		t.Fatal("expected signature mismatch error")
	}
}

func TestVerifySignature_badFormat(t *testing.T) {
	if err := governance.VerifySignature([]byte("x"), "noequalssign", "s"); err == nil {
		t.Fatal("expected format error")
	}
}

File: packages/sdk-ts/src/webhooks.test.ts
import { describe, expect, it } from "vitest";
import { verifyWebhookSignature } from "./webhooks.js";
import { createHmac } from "node:crypto";

const SECRET = "test-secret-abc123";
const PAYLOAD = JSON.stringify({ event: "policy.created" });

function sign(payload: string, secret: string): string {
  const mac = createHmac("sha256", secret).update(payload).digest("hex");
  return `sha256=${mac}`;
}

describe("verifyWebhookSignature", () => {
  it("returns true for a valid signature", async () => {
    const sig = sign(PAYLOAD, SECRET);
    const ok = await verifyWebhookSignature(PAYLOAD, sig, SECRET);
    expect(ok).toBe(true);
  });

  it("returns false for a tampered payload", async () => {
    const sig = sign(PAYLOAD, SECRET);
    const ok = await verifyWebhookSignature(PAYLOAD + "tampered", sig, SECRET);
    expect(ok).toBe(false);
  });

  it("returns false for a bad signature format", async () => {
    const ok = await verifyWebhookSignature(PAYLOAD, "md5=notvalid", SECRET);
    expect(ok).toBe(false);
  });

  it("returns false for wrong secret", async () => {
    const sig = sign(PAYLOAD, "wrong-secret");
    const ok = await verifyWebhookSignature(PAYLOAD, sig, SECRET);
    expect(ok).toBe(false);
  });
});
